spot_driver_cpp/src/api/default_robot_state_client.cpp

```diff
@@ -0,0 +1,382 @@
+    power_state.shore_power_state = robot_state.power_state().shore_power_state();
+    power_state.locomotion_charge_percentage = robot_state.power_state().locomotion_charge_percentage().value();
+    power_state.locomotion_estimated_runtime =
+        builtin_interfaces::build<builtin_interfaces::msg::Duration>()
+            .sec(robot_state.power_state().locomotion_estimated_runtime().seconds())
+            .nanosec(robot_state.power_state().locomotion_estimated_runtime().nanos());
+
+    return power_state;
+  }
+
+  return {};
+}
+
+std::optional<spot_msgs::msg::SystemFaultState> GetSystemFaultState(const ::bosdyn::api::RobotState& robot_state,
+                                                                    const google::protobuf::Duration& clock_skew) {
+  if (robot_state.has_system_fault_state()) {
+    spot_msgs::msg::SystemFaultState system_fault_state;
+    const auto create_fault_message = [&clock_skew](const ::bosdyn::api::SystemFault& fault) {
+      spot_msgs::msg::SystemFault fault_msg;
+      fault_msg.name = fault.name();
+      fault_msg.header.stamp = spot_ros2::applyClockSkew(fault.onset_timestamp(), clock_skew);
+      fault_msg.duration = builtin_interfaces::build<builtin_interfaces::msg::Duration>()
+                               .sec(fault.duration().seconds())
+                               .nanosec(fault.duration().nanos());
+      fault_msg.code = fault.code();
+      fault_msg.uid = fault.uid();
+      fault_msg.error_message = fault.error_message();
+      for (const auto& attr : fault.attributes()) {
+        fault_msg.attributes.push_back(attr);
+      }
+      fault_msg.severity = fault.severity();
+      return fault_msg;
+    };
+    for (const auto& fault : robot_state.system_fault_state().faults()) {
+      system_fault_state.faults.push_back(create_fault_message(fault));
+    }
+    for (const auto& fault : robot_state.system_fault_state().historical_faults()) {
+      system_fault_state.historical_faults.push_back(create_fault_message(fault));
+    }
+    return system_fault_state;
+  }
+  return {};
+}
+std::optional<bosdyn_msgs::msg::ManipulatorState> GetManipulatorState(const ::bosdyn::api::RobotState& robot_state) {
+  if (robot_state.has_manipulator_state()) {
+    bosdyn_msgs::msg::ManipulatorState manipulator_state;
+
+    manipulator_state.gripper_open_percentage = robot_state.manipulator_state().gripper_open_percentage();
+    manipulator_state.is_gripper_holding_item = robot_state.manipulator_state().is_gripper_holding_item();
+
+    manipulator_state.estimated_end_effector_force_in_hand.x =
+        robot_state.manipulator_state().estimated_end_effector_force_in_hand().x();
+    manipulator_state.estimated_end_effector_force_in_hand.y =
+        robot_state.manipulator_state().estimated_end_effector_force_in_hand().y();
+    manipulator_state.estimated_end_effector_force_in_hand.z =
+        robot_state.manipulator_state().estimated_end_effector_force_in_hand().z();
```

Comment:
Would it be worthwhile to add a few other conversion functions to avoid having to copy over individual fields for forces and velocities?

Reply:
I think so. @gremigi-bdai has generic conversion functions in another PR that I will cherry pick from